### PR TITLE
cross-test on a Windows host

### PR DESCRIPTION
**Pull request by RalfJung**

### CODE REVIEW

**Comment by RalfJung** [comment]
@CAD97 since you seem to use Windows, do you have any idea how to *unset* an environment variable in a batch script?

Somehow, batch scripts is an even worse language than bash scripts, and that really should be impossible.^^ There's a hole lot of code duplication here, but I don't dare even try factoring that into a function.

**Comment by RalfJung** [comment]
Also Cc @JOE1994

File: .appveyor.yml [hunk 1]
```diff
@@ -43,18 +42,36 @@ build_script:
...
     - set RUST_BACKTRACE=1
-    # Test miri
+    # Test host miri: 32bit Windows
+    - cargo miri setup
+    - set MIRI_SYSROOT=%USERPROFILE%\AppData\Local\rust-lang\miri\cache\HOST
+    - cargo test --release --all-features --locked
+    - cd test-cargo-miri
+    - '"C:\msys64\mingw64\bin\python3.exe" run-test.py'
+    - cd ..
+    - set MIRI_SYSROOT=
```
Comment (RalfJung) [inline .appveyor.yml]: This here seems to not have the effect of deleting the env var, even though that's what [the internet](https://stackoverflow.com/questions/7219672/how-to-undefine-a-ms-dos-variable) says I should do.

**Comment by CAD97** [comment]
In CMD, `set VAR=` is in fact the way to "undefine" an environment variable. Likely what's happening is that the executed command includes trailing whitespace for some reason; `set "VAR="` _should_ prevent this being an issue.

```cmd
C:\Users\CAD>if defined demo ( echo defined )

C:\Users\CAD>if not defined demo ( echo not defined )
not defined

C:\Users\CAD>set demo=demo

C:\Users\CAD>if defined demo ( echo defined )
defined

C:\Users\CAD>if not defined demo ( echo not defined )

C:\Users\CAD>set demo=

C:\Users\CAD>if defined demo ( echo defined )

C:\Users\CAD>if not defined demo ( echo not defined )
not defined
```

I typically avoid command line scripting though, and I stick to Powershell when I have to.

**Comment by RalfJung** [comment]
>In CMD, set VAR= is in fact the way to "undefine" an environment variable. Likely what's happening is that the executed command includes trailing whitespace for some reason; set "VAR=" should prevent this being an issue.

I'll try that, thanks!

> I typically avoid command line scripting though, and I stick to Powershell when I have to

Is there a way to do that with our AppVeyor scripting?

> (And typically, I try to treat "empty string" as unset because of this exact issue where unsetting a variable often times just sets it to an empty string, but it's still set.)

I'd rather not change the platform-independent Rust part because of the idiosyncrasies of a 30 year old scripting language...

**Comment by RalfJung** [comment]
It still says
```
[00:08:24] WARNING: MIRI_SYSROOT already set, not doing anything.
```
So, this hasn't worked. :/

**Comment by CAD97** [comment]
Ok, so here's something else to try:

<https://www.appveyor.com/docs/build-configuration/#interpreters-and-scripts>: you can use powershell scripts via `- ps: Powershell-Command`.

In powershell, the syntax would be

```powershell
PS C:\Users\CAD> Test-Path env:Demo
False
PS C:\Users\CAD> $env:Demo="demo"
PS C:\Users\CAD> Test-Path env:Demo
True
PS C:\Users\CAD> $env:Demo=$null
PS C:\Users\CAD> Test-Path env:Demo
False
```

**Comment by RalfJung** [comment]
But then I would have to port the entire thing to PowerShell, right? I never used that language and can't test this script locally, so I don't think that would work.

I was going to try and see if I can't use a bash script on Windows instead -- maybe AppVeyor has git-for-Windows installed which, from what I hear, comes with a bash. Then we could even share some code with the Travis side of things.

If that doesn't work, I think someone else would have to rewrite this for PowerShell.

**Comment by JOE1994** [comment]
> But then I would have to port the entire thing to PowerShell, right? I never used that language and can't test this script locally, so I don't think that would work.

[appVeyor allows us to choose between Command scripting and PowerShell scripting.](https://www.appveyor.com/docs/build-configuration/#interpreters-and-scripts)

Replacing the lines `- set "MIRI_SYSROOT="` to `- ps: $env:MIRI_SYSROOT = ""` solved the issue; 
environment variable gets undefined properly. ([Link to appveyor build of my fork](https://ci.appveyor.com/project/JOE1994/miri/builds/31641239#L1026)).

p.s. replacing the lines `- set "MIRI_SYSROOT="` to `- cmd: set "MIRI_SYSROOT="`doesn't solve the issue. ([Link to another appveyor build of my fork](https://ci.appveyor.com/project/JOE1994/miri?fullLog=true#L1028))

File: .appveyor.yml [hunk 2]
```diff
@@ -43,18 +42,36 @@ build_script:
...
     - set RUST_BACKTRACE=1
-    # Test miri
+    # Test host miri: 32bit Windows
+    - cargo miri setup
+    - set MIRI_SYSROOT=%USERPROFILE%\AppData\Local\rust-lang\miri\cache\HOST
+    - cargo test --release --all-features --locked
+    - cd test-cargo-miri
+    - '"C:\msys64\mingw64\bin\python3.exe" run-test.py'
+    - cd ..
+    - set "MIRI_SYSROOT="
```
Comment (JOE1994) [inline .appveyor.yml]:
```suggestion
    - ps: $env:MIRI_SYSROOT = ""
```

File: .appveyor.yml [hunk 3]
```diff
@@ -43,18 +42,36 @@ build_script:
...
+    # Test foreign miri: 64bit Linux
+    - cargo miri setup --target x86_64-unknown-linux-gnu
+    - set MIRI_SYSROOT=%USERPROFILE%\AppData\Local\rust-lang\miri\cache
+    - set MIRI_TEST_TARGET=x86_64-unknown-linux-gnu
+    - cargo test --release --all-features --locked
+    - cd test-cargo-miri
+    - '"C:\msys64\mingw64\bin\python3.exe" run-test.py'
+    - cd ..
+    - set "MIRI_SYSROOT="
```
Comment (JOE1994) [inline .appveyor.yml]:
```suggestion
    - ps: $env:MIRI_SYSROOT = ""
```

File: .appveyor.yml [hunk 4]
```diff
@@ -43,18 +42,36 @@ build_script:
...
+    # Test foreign miri: 64bit macOS
+    - cargo miri setup --target x86_64-apple-darwin
+    - set MIRI_SYSROOT=%USERPROFILE%\AppData\Local\rust-lang\miri\cache
+    - set MIRI_TEST_TARGET=x86_64-apple-darwin
     - cargo test --release --all-features --locked
-    # Test cargo integration
     - cd test-cargo-miri
     - '"C:\msys64\mingw64\bin\python3.exe" run-test.py'
+    - cd ..
+    - set "MIRI_SYSROOT="
```
Comment (JOE1994) [inline .appveyor.yml]:
```suggestion
    - ps: $env:MIRI_SYSROOT = ""
```

**Comment by RalfJung** [comment]
Yeah that seems to have worked, thanks @JOE1994! Now we are down to some individual tests failing:
```
failures:
    [ui] run-pass\current_dir.rs
    [ui] run-pass\fs.rs
    [ui] run-pass\libc.rs
```
My guess is that this is about directory separators.

**Comment by RalfJung** [comment]
Yeah, it tries to create `/tmp/miri_test_fs_file.txt`... after separator conversion, this becomes `\tmp\miri_test_fs_file.txt`, which of course does not work either.

**Comment by CAD97** [comment]
If it's being created from Rust, [`tempfile`](https://crates.io/crates/tempfile) is a generic way of handling transient temporary files. [`directories`](https://crates.io/crates/directories) offers ways of getting the system known folders, which doesn't seem to include a temp directory at the moment.

**Comment by RalfJung** [comment]
The problem is, this is running a linux target on a Windows host. So getting the system folders inside the interpreter is going to give the wrong result.

I think I am going to put the host temp dir into an env var in the test runner.

**Comment by RalfJung** [comment]
Looking good, isn't it? :D
@bors r+

**Comment by RalfJung** [comment]
Looks like travis has trouble reporting the status back? CI is green but bors/GH do not notice...

**Comment by RalfJung** [comment]
https://www.traviscistatus.com/ claims the issue is gone... let's see about that.
@bors retry

**Comment by RalfJung** [comment]
Hm, I just realized I should probably have waited for some feedback here... @oli-obk with this PR, the Miri shims that take file system paths replace `/` by `\` or vice versa when host and target OS directory separators differ (e.g. when executing `--target i686-unknown-linux-gnu` on a Windows host). This is the only good way I could think of to actually make things work, because the Linux target methods obviously expect `/` as directory separator. Does that seem reasonable?

**Pull request merged**